File: esl/simulation/agent_collection.py
from typing import Dict, Set, List

from esl.agent import Agent
from esl.simulation.identity import Identity


class AgentCollection(object):
    local_agents_: Dict[Identity, Agent]# Dict[Identity[Agent], Agent]
    global_agents_: Set[Identity]  # Identity[Agent]
    activated_: List[Identity] # Identity[Agent]
    deactivated_: List[Identity] # Identity[Agent]

    def __init__( self
                , local_agents: Dict[Identity, Agent] = None
                , global_agents: Set[Identity] = None
                ):
        if local_agents is None:
            local_agents = dict()
        if global_agents is None:
            global_agents = set()

        self.local_agents_ = local_agents
        self.global_agents_ = global_agents

    def activate(self, a: Agent):
        self.global_agents_.add(a.identifier)
        self.local_agents_[a.identifier] = a
        self.activated_.append(a)  # self.environment_.activate_agent(a)

    def deactivate(self, a: Agent):
        self.global_agents_.remove(a.identifier)
        del self.local_agents_[a.identifier]
        self.deactivated_.append(a)  # self.environment_.deactivate_agent(a)


    def __repr__(self):
        return f"AgentCollection(local_agents={self.local_agents_},global_agents={self.global_agents_})"

File: esl/simulation/agent_collection.cpp
/// \file   agent_collection.cpp
///
/// \brief
///
/// \authors    Maarten P. Scholl
/// \date       2019-04-27
/// \copyright  Copyright 2017-2019 The Institute for New Economic Thinking,
///             Oxford Martin School, University of Oxford
///
///             Licensed under the Apache License, Version 2.0 (the "License");
///             you may not use this file except in compliance with the License.
///             You may obtain a copy of the License at
///
///                 http://www.apache.org/licenses/LICENSE-2.0
///
///             Unless required by applicable law or agreed to in writing,
///             software distributed under the License is distributed on an "AS
///             IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
///             express or implied. See the License for the specific language
///             governing permissions and limitations under the License.
///
///             You may obtain instructions to fulfill the attribution
///             requirements in CITATION.cff
///
#include <esl/simulation/agent_collection.hpp>

#include <esl/agent.hpp>

#include <esl/computation/environment.hpp>


namespace esl::simulation {

    agent_collection::agent_collection(computation::environment &environment_)
    : environment_(environment_)
    {

    }

    void agent_collection::activate(std::shared_ptr<agent> a)
    {
        global_agents_.insert(a->identifier);
        local_agents_.insert({a->identifier, a});
        environment_.activate_agent(a->identifier);
    }

    void agent_collection::deactivate(std::shared_ptr<agent> a)
    {
        global_agents_.erase(a->identifier);
        local_agents_.erase(a->identifier);
        environment_.deactivate_agent(a->identifier);
    }

    std::shared_ptr<agent>
        agent_collection::operator[](const identity<agent> &a)
    {
        return local_agents_[a];
    }
}
